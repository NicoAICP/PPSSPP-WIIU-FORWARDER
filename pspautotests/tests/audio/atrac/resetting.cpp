#include "shared.h"

void testBufferInfo(const char *title, int atracID, int sample, bool withInfo, u8 *basePtr) {
	AtracResetBufferInfo info;
	memset(&info, 0xCC, sizeof(info));

	checkpoint("%s: %08x", title, sceAtracGetBufferInfoForResetting(atracID, sample, withInfo ? &info : NULL));
	schedfResetBuffer(info, basePtr);
}

extern "C" int main(int argc, char *argv[]) {
	Atrac3File at3("sample.at3");
	at3.Require();
	LoadAtrac();

	int ignore;
	int atracID;
	u16 data[16384];
	
	atracID = sceAtracSetHalfwayBufferAndGetID((u8 *)at3.Data(), at3.Size() / 2, at3.Size() / 2);
	checkpointNext("IDs:");
	testBufferInfo("  Unallocated", 4, 0, true, at3.Data());
	testBufferInfo("  Invalid", -1, 0, true, at3.Data());
	testBufferInfo("  Valid", atracID, 0, true, at3.Data());
	
	checkpointNext("Sample values:");
	const static int samples[] = {0, 1, 1679, 1680, 2046, 2047, 0x1000, 0x47ff, 0x10000, 247499, 247500, 247501, 0x100000, 0x1000000, 0x7FFFFFFF, -1, -2, -2048, -4096};
	for (size_t i = 0; i < ARRAY_SIZE(samples); ++i) {
		char temp[128];
		snprintf(temp, 128, "  %08x", samples[i]);
		testBufferInfo(temp, atracID, samples[i], true, at3.Data());
	}

	checkpointNext("Other states:");
	int withoutDataID = sceAtracGetAtracID(0x1000);
	testBufferInfo("  No data", withoutDataID, 0, true, at3.Data());
	sceAtracReleaseAtracID(withoutDataID);
	forceAtracState(atracID, 8);
	testBufferInfo("  State 8", atracID, 0, true, at3.Data());
	forceAtracState(atracID, 16);
	testBufferInfo("  State 16", atracID, 0, true, at3.Data());
	forceAtracState(atracID, 1);
	sceAtracReleaseAtracID(atracID);

	// Crashes.
	//checkpointNext("Buffer:");
	//testBufferInfo("  NULL", atracID, 0, false, at3.Data());

	at3.Reload("sample.at3");
	atracID = sceAtracSetDataAndGetID(at3.Data(), at3.Size());
	checkpointNext("Entire buffer:");
	testBufferInfo("  At start -> 0", atracID, 0, true, at3.Data());
	checkpoint("  Decode: %08x", sceAtracDecodeData(atracID, data, &ignore, &ignore, &ignore));
	testBufferInfo("  After decode 1 -> 0", atracID, 0, true, at3.Data());

	testBufferInfo("  After decode 1 -> 65536", atracID, 65536, true, at3.Data());
	checkpoint("  Decode: %08x", sceAtracDecodeData(atracID, data, &ignore, &ignore, &ignore));
	testBufferInfo("  After decode 2 -> 65536", atracID, 65536, true, at3.Data());

	testBufferInfo("  After decode 2 -> 196608", atracID, 196608, true, at3.Data());
	checkpoint("  Decode: %08x", sceAtracDecodeData(atracID, data, &ignore, &ignore, &ignore));
	testBufferInfo("  After decode 3 -> 196608", atracID, 196608, true, at3.Data());

	sceAtracReleaseAtracID(atracID);
	at3.Reload("sample.at3");
	atracID = sceAtracSetHalfwayBufferAndGetID((u8 *)at3.Data(), at3.Size() / 2, at3.Size());
	checkpointNext("Half buffer:");
	testBufferInfo("  At start -> 0", atracID, 0, true, at3.Data());
	checkpoint("  Decode: %08x", sceAtracDecodeData(atracID, data, &ignore, &ignore, &ignore));
	testBufferInfo("  After decode 1 -> 0", atracID, 0, true, at3.Data());

	testBufferInfo("  After decode 1 -> 65536", atracID, 65536, true, at3.Data());
	checkpoint("  Decode: %08x", sceAtracDecodeData(atracID, data, &ignore, &ignore, &ignore));
	testBufferInfo("  After decode 2 -> 65536", atracID, 65536, true, at3.Data());

	testBufferInfo("  After decode 2 -> 196608", atracID, 196608, true, at3.Data());
	checkpoint("  Decode: %08x", sceAtracDecodeData(atracID, data, &ignore, &ignore, &ignore));
	testBufferInfo("  After decode 3 -> 196608", atracID, 196608, true, at3.Data());

	sceAtracReleaseAtracID(atracID);
	at3.Reload("sample.at3");
	atracID = sceAtracSetHalfwayBufferAndGetID((u8 *)at3.Data(), at3.Size() / 2, at3.Size() / 2);
	checkpointNext("Streamed buffer:");
	testBufferInfo("  At start -> 0", atracID, 0, true, at3.Data());
	checkpoint("  Decode: %08x", sceAtracDecodeData(atracID, data, &ignore, &ignore, &ignore));
	testBufferInfo("  After decode 1 -> 0", atracID, 0, true, at3.Data());

	testBufferInfo("  After decode 1 -> 65536", atracID, 65536, true, at3.Data());
	checkpoint("  Decode: %08x", sceAtracDecodeData(atracID, data, &ignore, &ignore, &ignore));
	testBufferInfo("  After decode 2 -> 65536", atracID, 65536, true, at3.Data());

	testBufferInfo("  After decode 2 -> 196608", atracID, 196608, true, at3.Data());
	checkpoint("  Decode: %08x", sceAtracDecodeData(atracID, data, &ignore, &ignore, &ignore));
	testBufferInfo("  After decode 3 -> 196608", atracID, 196608, true, at3.Data());

	return 0;
}
