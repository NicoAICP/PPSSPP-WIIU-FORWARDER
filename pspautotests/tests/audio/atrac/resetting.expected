[x] IDs:
[x]   Unallocated: 80630005
   #1: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
   #2: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
[x]   Invalid: 80630005
   #1: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
   #2: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
[x]   Valid: 00000000
   #1: write=p+0x0, writable=00005998, min=000002f0, file=00000060
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000

[x] Sample values:
[x]   00000000: 00000000
   #1: write=p+0x0, writable=00005998, min=000002f0, file=00000060
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   00000001: 00000000
   #1: write=p+0x0, writable=00005998, min=000002f0, file=00000060
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   0000068f: 00000000
   #1: write=p+0x0, writable=00005998, min=000002f0, file=00000060
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   00000690: 00000000
   #1: write=p+0x0, writable=00005998, min=00000468, file=00000060
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   000007fe: 00000000
   #1: write=p+0x0, writable=00005998, min=00000468, file=00000060
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   000007ff: 00000000
   #1: write=p+0x0, writable=00005998, min=00000468, file=00000060
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   00001000: 00000000
   #1: write=p+0x0, writable=00005998, min=000002f0, file=00000350
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   000047ff: 00000000
   #1: write=p+0x0, writable=00005998, min=00000468, file=00000c20
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   00010000: 00000000
   #1: write=p+0x0, writable=00005998, min=000002f0, file=00002f60
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   0003c6cb: 00000000
   #1: write=p+0x0, writable=00000468, min=00000468, file=0000b0a0
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   0003c6cc: 00000000
   #1: write=p+0x0, writable=00000468, min=00000468, file=0000b0a0
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   0003c6cd: 80630015
   #1: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
   #2: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
[x]   00100000: 80630015
   #1: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
   #2: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
[x]   01000000: 80630015
   #1: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
   #2: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
[x]   7fffffff: 80630015
   #1: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
   #2: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
[x]   ffffffff: 00000000
   #1: write=p+0x0, writable=00005998, min=00000468, file=00000060
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   fffffffe: 00000000
   #1: write=p+0x0, writable=00005998, min=00000468, file=00000060
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   fffff800: 00000000
   #1: write=p+0x0, writable=00005998, min=000002f0, file=fffffee8
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   fffff000: 80630015
   #1: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
   #2: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc

[x] Other states:
[x]   No data: 80630010
   #1: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
   #2: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
[x]   State 8: 80630031
   #1: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
   #2: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
[x]   State 16: 80630040
   #1: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc
   #2: write=0xcccccccc, writable=cccccccc, min=cccccccc, file=cccccccc

[x] Entire buffer:
[x]   At start -> 0: 00000000
   #1: write=p+0x0, writable=00000000, min=00000000, file=00000000
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[r]   Decode: 00000000
[x]   After decode 1 -> 0: 00000000
   #1: write=p+0x0, writable=00000000, min=00000000, file=00000000
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   After decode 1 -> 65536: 00000000
   #1: write=p+0x0, writable=00000000, min=00000000, file=00000000
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[r]   Decode: 00000000
[x]   After decode 2 -> 65536: 00000000
   #1: write=p+0x0, writable=00000000, min=00000000, file=00000000
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   After decode 2 -> 196608: 00000000
   #1: write=p+0x0, writable=00000000, min=00000000, file=00000000
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[r]   Decode: 00000000
[x]   After decode 3 -> 196608: 00000000
   #1: write=p+0x0, writable=00000000, min=00000000, file=00000000
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000

[x] Half buffer:
[x]   At start -> 0: 00000000
   #1: write=p+0x5a84, writable=00005a84, min=00000000, file=00005a84
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[r]   Decode: 00000000
[x]   After decode 1 -> 0: 00000000
   #1: write=p+0x5a84, writable=00005a84, min=00000000, file=00005a84
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   After decode 1 -> 65536: 00000000
   #1: write=p+0x5a84, writable=00005a84, min=00000000, file=00005a84
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[r]   Decode: 00000000
[x]   After decode 2 -> 65536: 00000000
   #1: write=p+0x5a84, writable=00005a84, min=00000000, file=00005a84
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   After decode 2 -> 196608: 00000000
   #1: write=p+0x5a84, writable=00005a84, min=000035cc, file=00005a84
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[r]   Decode: 00000000
[x]   After decode 3 -> 196608: 00000000
   #1: write=p+0x5a84, writable=00005a84, min=000035cc, file=00005a84
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000

[x] Streamed buffer:
[x]   At start -> 0: 00000000
   #1: write=p+0x0, writable=00005998, min=000002f0, file=00000060
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[r]   Decode: 00000000
[x]   After decode 1 -> 0: 00000000
   #1: write=p+0x0, writable=00005998, min=000002f0, file=00000060
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   After decode 1 -> 65536: 00000000
   #1: write=p+0x0, writable=00005998, min=000002f0, file=00002f60
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[r]   Decode: 00000000
[x]   After decode 2 -> 65536: 00000000
   #1: write=p+0x0, writable=00005998, min=000002f0, file=00002f60
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[x]   After decode 2 -> 196608: 00000000
   #1: write=p+0x0, writable=000027a8, min=000002f0, file=00008d60
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
[r]   Decode: 00000000
[x]   After decode 3 -> 196608: 00000000
   #1: write=p+0x0, writable=000027a8, min=000002f0, file=00008d60
   #2: write=p+0x0, writable=00000000, min=00000000, file=00000000
